### ISSUE TITLE
Don't throw errors in TypeScript

### ISSUE THREAD

**Issue by robertbastian** [rust-diplomat/diplomat#554]
TS doesn't seem to have a way to declare that methods throw errors. The best it can do is:
```typescript
/// Throws `ErrorEnum`
static newFailingFoo(): ResultOpaque | undefined;
```

However, it has union types, so we can do something much better:
```typescript
static newFailingFoo(): ResultOpaque | ErrorEnum;
```

In this case, the `ErrorEnum` would be `return`ed, not `throw`n. This means, however, that JS without types also has to do this.

Thoughts?

**Comment by CBenoit** [comment]
I’ve heard of this pattern in TypeScript before, and actually like it better as long as TypeScript is used, but it does not sounds very nice when using JavaScript alone as there is no type checking enforcing the check anymore… maybe not worse than usual JavaScript though?

**Comment by robertbastian** [comment]
I've just played around with this a bit, and it's actually annoying to use on the call side in TS, as there's not really any support to destructure union types. `try/catch` is easier to use.

**Comment by CBenoit** [comment]
And no `?` operator like in Rust for propagating the errors easily either

**Comment by Manishearth** [comment]
Yeah I recall briefly looking into this and deciding it's not worth it. However if someone comes up with a good way to do this we should!

**Comment by ambiguousname** [comment]
Looking into how we could annotate with typescript, one solution is the [`never`](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#the-never-type) type:

```ts
static newFailingFoo(): ResultOpaque | never;
```

`never` by itself is kind of meaningless, so we could define a custom type to show what kind of error it is:

```ts
type ThrowsErrorEnum = never;
static newFailingFoo(): ResultOpaque | ThrowsErrorEnum;
```

**Comment by robertbastian** [comment]
This is also outside the type system, so not sure if it's an improvement.
